File: src/running_stats.cpp
/* Implementation of class RunningStats
   One-pass algorithm for mean and variance.
   Chris Toney <chris.toney at usda.gov> */

#include "running_stats.h"

RunningStats::RunningStats():
    m_na_rm(true), m_count(0) {}

RunningStats::RunningStats(bool na_rm):
    m_na_rm(na_rm), m_count(0) {}

void RunningStats::update(const Rcpp::NumericVector& newvalues) {
    for (auto const& i : newvalues) {
        if (m_na_rm) {
            if (Rcpp::NumericVector::is_na(i))
                continue;
        }
        m_count += 1;
        if (m_count == 1) {
            m_mean = m_min = m_max = m_sum = i;
            m_M2 = 0.0;
        }
        else {
            const double delta = i - m_mean;
            m_mean += (delta / m_count);
            const double delta2 = i - m_mean;
            m_M2 += (delta * delta2);
            if (i < m_min)
                m_min = i;
            else if (i > m_max)
                m_max = i;
            m_sum += i;
        }
    }
}

void RunningStats::reset() {
    m_count = 0;
}

double RunningStats::get_count() const {
    // return as double in R (no native int64)
    return static_cast<double>(m_count);
}

double RunningStats::get_mean() const {
    if (m_count > 0)
        return m_mean;
    else
        return NA_REAL;
}

// From R help for min/max:
// "The minimum and maximum of a numeric empty set are ‘+Inf’ and
// ‘-Inf’ (in this order!) which ensures _transitivity_, e.g.,
// ‘min(x1, min(x2)) == min(x1, x2)’."
double RunningStats::get_min() const {
    if (Rcpp::NumericVector::is_na(m_sum))
        return NA_REAL;

    if (m_count > 0)
        return m_min;
    else
        return R_PosInf;
}

double RunningStats::get_max() const {
    if (Rcpp::NumericVector::is_na(m_sum))
        return NA_REAL;

    if (m_count > 0)
        return m_max;
    else
        return R_NegInf;
}

double RunningStats::get_sum() const {
    if (m_count > 0)
        return m_sum;
    else
        return 0;
}

double RunningStats::get_var() const {
    if (m_count < 2)
        return NA_REAL;
    else
        return (m_M2 / (m_count - 1));
}

double RunningStats::get_sd() const {
    if (m_count < 2)
        return NA_REAL;
    else
        return sqrt(m_M2 / (m_count - 1));
}

void RunningStats::show() const {
    Rcpp::Rcout << "C++ object of class RunningStats" << std::endl;
    Rcpp::Rcout << " Number of values: " << get_count() << std::endl;
}

RCPP_MODULE(mod_running_stats) {
    Rcpp::class_<RunningStats>("RunningStats")

    .constructor
        ("Default constructor initialized with na_rm = TRUE.")
    .constructor<bool>
        ("Initialize with na_rm = TRUE or FALSE")

    .method("update", &RunningStats::update,
        "Add new values from a numeric vector")
    .method("reset", &RunningStats::reset,
        "Reset the data stream to count = 0")

    .const_method("get_count", &RunningStats::get_count,
        "Return the count of values currently in the stream")
    .const_method("get_mean", &RunningStats::get_mean,
        "Return the mean of the values currently in the stream")
    .const_method("get_min", &RunningStats::get_min,
        "Return the minimum value currently in the stream")
    .const_method("get_max", &RunningStats::get_max,
        "Return the maximum value currently in the stream")
    .const_method("get_sum", &RunningStats::get_sum,
        "Return the sum of values currently in the stream")
    .const_method("get_var", &RunningStats::get_var,
        "Return the variance of the values currently in the stream")
    .const_method("get_sd", &RunningStats::get_sd,
        "Return standard deviation of the values currently in the stream")
    .const_method("show", &RunningStats::show,
        "S4 show()")
    ;
}

File: R/running_stats.R
#' @name RunningStats-class
#'
#' @aliases
#' Rcpp_RunningStats Rcpp_RunningStats-class RunningStats
#'
#' @title Class to calculate mean and variance in one pass
#'
#' @description
#' `RunningStats` computes summary statistics on a data stream efficiently.
#' Mean and variance are calculated with Welford's online algorithm
#' (\url{https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance}).
#' The min, max, sum and count are also tracked. The input data values are not 
#' stored in memory, so this class can be used to compute statistics for very 
#' large data streams.
#'
#' @param na_rm Logical scalar. `TRUE` to remove `NA` from the input data or
#' `FALSE` to retain `NA` (defaults to `TRUE`).
#' @returns An object of class `RunningStats`. A `RunningStats` object 
#' maintains the current minimum, maximum, mean, variance, sum and count of 
#' values that have been read from the stream. It can be updated repeatedly 
#' with new values (i.e., chunks of data read from the input stream), but its 
#' memory footprint is negligible. Class methods for updating with new values 
#' and retrieving current values of statistics are described in Details.
#' `RunningStats` is a C++ class exposed directly to R (via
#' `RCPP_EXPOSED_CLASS`). Methods of the class are accessed in R using the `$`
#' operator.
#'
#' @note
#' The intended use is computing summary statistics for specific subsets or
#' zones of a raster that could be defined in various ways and are generally
#' not contiguous. The algorithm as implemented here incurs the cost of
#' floating point division for each new value updated (i.e., per pixel), but is
#' reasonably efficient for the use case. Note that GDAL internally uses an
#' optimized version of Welford's algorithm to compute raster statistics as
#' described in detail by Rouault, 2016
#' (\url{https://github.com/OSGeo/gdal/blob/master/gcore/statistics.txt}).
#' The class method `GDALRaster$getStatistics()` is a GDAL API wrapper that
#' computes statistics for a whole raster band.
#'
#' @section Usage:
#' \preformatted{
#' rs <- new(RunningStats, na_rm)
#'
#' ## Methods (see Details)
#' rs$update(newvalues)
#' rs$get_count()
#' rs$get_mean()
#' rs$get_min()
#' rs$get_max()
#' rs$get_sum()
#' rs$get_var()
#' rs$get_sd()
#' rs$reset()
#' }
#'
#' @section Details:
#'
#' \code{new(RunningStats, na_rm)}
#' Constructor. Returns an object of class \code{RunningStats}.
#'
#' \code{$update(newvalues)}
#' Updates the `RunningStats` object with a numeric vector of `newvalues` 
#' (i.e., a chunk of values from the data stream). No return value, called 
#' for side effects.
#'
#' \code{$get_count()}
#' Returns the count of values received from the data stream.
#'
#' \code{$get_mean()}
#' Returns the mean of values received from the data stream.
#'
#' \code{$get_min()}
#' Returns the minimum value received from the data stream.
#'
#' \code{$get_max()}
#' Returns the maximum value received from the data stream.
#'
#' \code{$get_sum()}
#' Returns the sum of values received from the data stream.
#'
#' \code{$get_var()}
#' Returns the variance of values from the data stream 
#' (denominator n - 1).
#'
#' \code{$get_sd()}
#' Returns the standard deviation of values from the data stream
#' (denominator n - 1).
#'
#' \code{$reset()}
#' Clears the \code{RunningStats} object to its initialized state (count = 0).
#' No return value, called for side effects.
#' 
#' @examples
#' set.seed(42)
#'
#' rs <- new(RunningStats, na_rm=TRUE)
#' chunk <- runif(1000)
#' rs$update(chunk)
#' object.size(rs)
#'
#' rs$get_count()
#' length(chunk)
#'
#' rs$get_mean()
#' mean(chunk)
#'
#' rs$get_min()
#' min(chunk)
#'
#' rs$get_max()
#' max(chunk)
#'
#' rs$get_var()
#' var(chunk)
#'
#' rs$get_sd()
#' sd(chunk)
#' 
#' \donttest{
#' ## 10^9 values read in 10,000 chunks
#' ## should take under 1 minute on most PC hardware
#' for (i in 1:1e4) {
#'   chunk <- runif(1e5)
#'   rs$update(chunk)
#' }
#' rs$get_count()
#' rs$get_mean()
#' rs$get_var()
#'
#' object.size(rs)
#' }
NULL

Rcpp::loadModule("mod_running_stats", TRUE)
